internal/repository/indexmap.go

```diff
@@ -0,0 +1,140 @@
+package repository
+
+import (
+	"encoding/binary"
+
+	"github.com/restic/restic/internal/restic"
+)
+
+// An indexMap is a chained hash table that maps blob IDs to indexEntries.
+// It allows storing multiple entries with the same key.
+//
+// A chained table has less memory overhead than a probing one (like the
+// standard map), because only a pointer array needs to be resized when the
+// table grows.
+type indexMap struct {
+	// The number of buckets is always a power of two and never zero.
+	buckets    []*indexEntry
+	numentries uint
+
+	free *indexEntry // Free list.
+}
+
+const (
+	growthFactor = 2 // Must be a power of 2.
+	maxLoad      = 4 // Max. number of entries per bucket.
+)
+
+// add inserts an indexEntry for the given arguments into the map,
+// using id as the key.
+func (m *indexMap) add(id restic.ID, packIdx int, offset, length uint32) {
+	switch {
+	case m.numentries == 0: // Lazy initialization.
+		m.buckets = make([]*indexEntry, 64)
+	case m.numentries >= maxLoad*uint(len(m.buckets)):
+		m.grow()
+	}
+
+	h := m.hash(id)
+	e := m.newEntry()
+	e.id = id
+	e.next = m.buckets[h]
+	e.packIndex = packIdx
+	e.offset = offset
+	e.length = length
+
+	m.buckets[h] = e
+	m.numentries++
+}
+
+// foreach calls fn for all entries in the map, until fn returns false.
+func (m *indexMap) foreach(fn func(*indexEntry) bool) {
+	for _, e := range m.buckets {
+		for e != nil {
+			if !fn(e) {
+				return
+			}
+			e = e.next
+		}
+	}
+}
+
+// foreachWithID calls fn for all entries with the given id.
+func (m *indexMap) foreachWithID(id restic.ID, fn func(*indexEntry)) {
+	if len(m.buckets) == 0 {
+		return
+	}
+
+	h := m.hash(id)
+	for e := m.buckets[h]; e != nil; e = e.next {
+		if e.id != id {
+			continue
+		}
+		fn(e)
+	}
+}
+
+// get returns the first entry for the given id.
+func (m *indexMap) get(id restic.ID) *indexEntry {
+	if len(m.buckets) == 0 {
+		return nil
+	}
+
+	h := m.hash(id)
+	for e := m.buckets[h]; e != nil; e = e.next {
+		if e.id == id {
+			return e
+		}
+	}
+	return nil
+}
+
+func (m *indexMap) grow() {
+	old := m.buckets
+	m.buckets = make([]*indexEntry, growthFactor*len(m.buckets))
+
+	for _, e := range old {
+		for e != nil {
+			h := m.hash(e.id)
+			next := e.next
+			e.next = m.buckets[h]
+			m.buckets[h] = e
+			e = next
+		}
+	}
+}
+
+func (m *indexMap) hash(id restic.ID) uint {
+	h := uint(binary.LittleEndian.Uint64(id[:8]))
```

Comment:
Note that id already is a SHA256 hash. It should pretty hard to create lot's of contents efficiently with identical first bytes. If it was easy, bitcoin would be worthless immediately.

Of course you could "manipulate" your date (e.g. index files) in some sense such that you do not save SHA256 hashes. However then you have restic processing garbage (and it will abort sooner or later) and hopefully don't care about performance :wink:

Reply:
Thinking out loud again.

All hashes are susceptible to birthday attacks: you can find a collision for the first *k* bits in 2^(*k*/2) time. To turn that into a hash flooding attack, you'd need to generate *n* × 2^(*k*/2) colliding strings for a chain of length *n*, each of length lg(*n*) - 1 bits, on average. To create a single linked list (*n* = 2^*k*), that comes down to (*k* - 1) 2^(1.5 *k*) or 23 × 2^36 work for 24 bits of hash. That's 1.47 TiB of data, disregarding the fact that only 1/8 bit strings are also possible byte strings.

Reply:
> Thinking out loud again.
> 
> All hashes are susceptible to birthday attacks: you can find a collision for the first _k_ bits in 2^(_k_/2) time. To turn that into a hash flooding attack, you'd need to generate _n_ × 2^(_k_/2) colliding strings for a chain of length _n_, each of length lg(_n_) - 1 bits, on average. To create a single linked list (_n_ = 2^_k_), that comes down to (_k_ - 1) 2^(1.5 _k_) or 23 × 2^36 work for 24 bits of hash. That's 1.47 TiB of data, disregarding the fact that only 1/8 bit strings are also possible byte strings.

I couldn't follow your computations without thinking in more depth about it. 
However I have the feeling that your argumentation may be wrong: When you argue with birthday attacks, you calculate the work to get an arbitrary hash collusion, i.e. two pieces of data with identical hashes. When looking for the third (or more) data piece, you are no longer free about which hash this should have, but you must find some data with that given hash. So finding lots of data with arbitrary identical hash is almost as hard as finding lots of data with a given hash.

Reply:
Then again, even for very small *k* it's possible to create a significant slowdown, esp. if you can generate multiple chains, which may be easier. I'll put in a randomized xxhash.

Reply:
Following the advice [here](https://crypto.stackexchange.com/q/6408), I've switched to SipHash instead. Maybe xxhash with SHA-256 inputs is safe, but I don't want to be the roll-your-own-crypto guy. My benchmarks show that siphash.Hash is only about 1.5× slower than xxhash.Sum64. This does have the effect of slowing some operations down compared to current master, but not by a great deal.

    PackerManager-8                             181ms ± 2%     179ms ± 1%   -1.25%  (p=0.005 n=10+10)
    DecodeIndex-8                              13.6µs ± 2%    18.1µs ± 3%  +32.46%  (p=0.000 n=10+10)
    IndexHasUnknown-8                          43.4ns ± 5%    52.3ns ±17%  +20.43%  (p=0.000 n=10+10)
    IndexHasKnown-8                            49.2ns ± 3%    50.1ns ± 8%     ~     (p=0.088 n=10+9)
    IndexAlloc-8                                809ms ± 2%     576ms ± 2%  -28.79%  (p=0.000 n=9+10)
    MasterIndexLookupSingleIndex-8              122ns ± 0%     152ns ± 4%  +25.00%  (p=0.000 n=6+10)
    MasterIndexLookupMultipleIndex-8            379ns ± 2%     445ns ±10%  +17.38%  (p=0.000 n=10+10)
    MasterIndexLookupSingleIndexUnknown-8      68.9ns ± 4%    70.9ns ±11%     ~     (p=0.304 n=9+10)
    MasterIndexLookupMultipleIndexUnknown-8     317ns ± 1%     347ns ± 4%   +9.28%  (p=0.000 n=9+9)
    SaveAndEncrypt-8                            101ns ± 2%      95ns ± 0%   -5.35%  (p=0.000 n=10+8)
    LoadTree-8                                  350µs ±12%     452µs ± 2%  +29.36%  (p=0.000 n=10+10)
    LoadBlob-8                                 5.78ms ± 2%    5.66ms ± 0%   -2.22%  (p=0.000 n=10+9)
    LoadIndex-8                                32.8ms ± 2%    31.7ms ± 2%   -3.43%  (p=0.000 n=10+10)
    
    name                                     old speed      new speed      delta
    PackerManager-8                           291MB/s ± 2%   295MB/s ± 1%   +1.26%  (p=0.005 n=10+10)
    SaveAndEncrypt-8                         41.7TB/s ± 2%  44.0TB/s ± 1%   +5.49%  (p=0.000 n=10+9)
    LoadBlob-8                                173MB/s ± 2%   177MB/s ± 0%   +2.25%  (p=0.000 n=10+9)
    
    name                                     old alloc/op   new alloc/op   delta
    PackerManager-8                            91.5kB ± 0%    91.5kB ± 0%     ~     (p=0.839 n=10+10)
    IndexAlloc-8                                786MB ± 0%     400MB ± 0%  -49.06%  (p=0.000 n=10+10)
    SaveAndEncrypt-8                            1.00B ± 0%     1.00B ± 0%     ~     (all equal)
    
    name                                     old allocs/op  new allocs/op  delta
    PackerManager-8                             1.41k ± 0%     1.41k ± 0%   -0.06%  (p=0.040 n=10+10)
    IndexAlloc-8                                 977k ± 0%     1116k ± 0%  +14.28%  (p=0.000 n=10+10)
    SaveAndEncrypt-8                             0.00           0.00          ~     (all equal)

Reply:
> When you argue with birthday attacks, you calculate the work to get an arbitrary hash collusion, i.e. two pieces of data with identical hashes. When looking for the third (or more) data piece, you are no longer free about which hash this should have, but you must find some data with that given hash. So finding lots of data with arbitrary identical hash is almost as hard as finding lots of data with an a priori given hash.

Sorry, I hadn't seen this message when I posted the previous comments.

Regardless of my previous reasoning, which may well be wrong, you don't even need to create such a long chain. You can trivially generate multiple chains of a few hundred blobs by looking for collisions in the upper bits of SHA-256. If you pad those with a big enough blob to meet minPackSize, you can generate an index that is two orders of magnitude slower than anything else.

Reply:
Go 1.14 has a new package that looks like a perfect match: https://golang.org/pkg/hash/maphash/ . However, it's not available on older Go versions.

Reply:
Despite maphash being 10-17% faster than siphash on its own, operations using it become slower. I removed the parallel grow to get rid of copying and ensured that I did not introduce any allocations in the hot paths, but I get these benchmark results:

```
name                                     old time/op    new time/op    delta
PackerManager-8                             179ms ± 1%     179ms ± 1%     ~     (p=0.190 n=10+10)
DecodeIndex-8                              18.1µs ± 3%    16.0µs ± 4%  -11.70%  (p=0.000 n=10+10)
IndexHasUnknown-8                          52.3ns ±17%    55.5ns ±13%     ~     (p=0.324 n=10+10)
IndexHasKnown-8                            50.1ns ± 8%    54.5ns ±11%   +8.78%  (p=0.000 n=9+10)
IndexAlloc-8                                576ms ± 2%     756ms ± 4%  +31.18%  (p=0.000 n=10+10)
MasterIndexLookupSingleIndex-8              152ns ± 4%     158ns ± 7%     ~     (p=0.075 n=10+10)
MasterIndexLookupMultipleIndex-8            445ns ±10%     469ns ±10%     ~     (p=0.105 n=10+10)
MasterIndexLookupSingleIndexUnknown-8      70.9ns ±11%    76.5ns ± 7%   +7.87%  (p=0.007 n=10+9)
MasterIndexLookupMultipleIndexUnknown-8     347ns ± 4%     378ns ± 7%   +9.10%  (p=0.000 n=9+10)
SaveAndEncrypt-8                           95.2ns ± 0%    99.9ns ± 0%   +4.91%  (p=0.000 n=8+7)
LoadTree-8                                  452µs ± 2%     491µs ± 3%   +8.63%  (p=0.000 n=10+10)
LoadBlob-8                                 5.66ms ± 0%    5.66ms ± 0%   +0.10%  (p=0.036 n=9+8)
LoadIndex-8                                31.7ms ± 2%    32.0ms ± 3%     ~     (p=0.243 n=10+9)
```